onnxruntime/python/tools/kernel_explorer/kernels/gemm_test.py

```diff
@@ -33,47 +24,81 @@ def _test_gemm(func, dtype: str, m: int, n: int, k: int, transa=False, transb=Fa
     np.random.seed(0)
     a = (np.random.rand(*a_shape) * 2 - 1).astype(dtype)
     b = (np.random.rand(*b_shape) * 2 - 1).astype(dtype)
+    ref_c = (a.T if transa else a) @ (b.T if transb else b)
 
     my_c = np.zeros((m, n), dtype=dtype)
     dev_a = ke.DeviceArray(a)
     dev_b = ke.DeviceArray(b)
     dev_c = ke.DeviceArray(my_c)
 
-    f = getattr(ke, func)
-
     opa = ke.blas_op.T if transa else ke.blas_op.N
     opb = ke.blas_op.T if transb else ke.blas_op.N
     lda = a_shape[1]
     ldb = b_shape[1]
     alpha = 1.0
     beta = 0.0
-    my_gemm = f(opa, opb, m, n, k, alpha, dev_a, lda, dev_b, ldb, beta, dev_c, n)
-    my_gemm.Run()
-    dev_c.UpdateHostNumpyArray()
+    my_gemm = func(opa, opb, m, n, k, alpha, dev_a, lda, dev_b, ldb, beta, dev_c, n)
 
-    rtol = 1e-3 if dtype == "float16" else 1e-5
-    atol = 1e-3 if dtype == "float16" else 1e-5
+    failures = {}
+    for impl in my_gemm.ListImpls():
+        if not my_gemm.SelectImpl(impl):
+            continue
 
-    print(a, b)
-    ref_c = (a.T if transa else a) @ (b.T if transb else b)
-    np.testing.assert_allclose(ref_c, my_c, rtol=rtol, atol=atol)
+        my_gemm.Run()
+        dev_c.UpdateHostNumpyArray()
+
+        rtol = 1e-3 if dtype == "float16" else 1e-5
+        atol = 1e-3 if dtype == "float16" else 1e-5
+
+        try:
+            np.testing.assert_allclose(ref_c, my_c, rtol=rtol, atol=atol)
+        except Exception as err:
+            failures[impl] = str(err)
 
+    if failures:
+        raise Exception(failures)
 
-dtypes = ["float32", "float16"]
-transabs = product([True, False], repeat=2)
-basic_sizes = product([1, 3, 4, 16, 127, 128, 129, 133, 1024], repeat=3)
+
+def get_basic_cases():
+    dtypes = ["float32", "float16"]
+    transabs = product([True, False], repeat=2)
+    basic_sizes = product([1, 3, 4, 16, 127, 128, 129, 133, 1024], repeat=3)
+    return list(product(dtypes, basic_sizes, transabs))
+
+
+def get_bert_cases():
+    dtypes = ["float32", "float16"]
+    transabs = [(False, False)]
+    bert_sizes = [
+        # m, n, k
+        (384, 64, 384),
+        (384, 768, 768),
+        (384, 3072, 768),
+        (384, 1024, 1024),
+        (384, 4096, 1024),
+        (384, 768, 3072),
+        (384, 1024, 4096),
+    ]
+    return list(product(dtypes, bert_sizes, transabs))
 
 
 @pytest.mark.parametrize(
     "dtype, size, transab",
-    product(dtypes, basic_sizes, transabs),
+    get_basic_cases() + get_bert_cases(),
 )
-def test_gemm_all_cases(dtype, size, transab):
-    for f in dtype_to_funcs(dtype):
-        _test_gemm(f, dtype, *size, *transab)
+def test_rocblas_gemm_all_cases(dtype, size, transab):
+    _test_gemm(getattr(ke, "RocblasGemm_" + dtype_to_suffix(dtype)), dtype, *size, *transab)
 
 
-def profile_gemm_func(func, dtype, m, n, k):
+@pytest.mark.parametrize(
+    "dtype, size, transab",
+    get_bert_cases(),
+)
+def test_ck_gemm_bert_cases(dtype, size, transab):
```

Comment:
What are the reasons to exclude basic cases here?

Reply:
For the basic test cases, counting all those combinations reach up to ~5k test cases. For CK's gemm with client api, each gemm will have ~20 impls for each. Then we will reach 100k test runs (implicit, hidden behind the `for` loop) for it.

Reply:
So we have 20x more kernels to test. Looks like the number of test cases is linear to L^3 = 9^3=729, where L is the size of array [1, 3, 4, 16, 127, 128, 129, 133, 1024], if we reduce L to 4, we have 4^3 = 64, which is a reduction of test cases of 10x to accommodate 20x more kernels. There might be other ways to reduce the number of test cases at your disposal. 

I think it is good to have these tests, they are more of irregular sizes and complementary to bert test cases, which are of more regular sizes.